### DISCUSSION TITLE
reducing-latency-by-pre-building-singletons-in-asp-net-core/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#3] [Announcements]
# reducing-latency-by-pre-building-singletons-in-asp-net-core/

In this post I show a startup task that pre-builds all the services registered in the DI container to reduce the duration of the first request to the app.

https://andrewlock.net/reducing-latency-by-pre-building-singletons-in-asp-net-core/

**Comment by trontronicent** [comment]
I know its been a while since release of this post, but anyway, its gold! My assumption, why even scoped services profit from it: Its due to the nature of the factory pattern used by many internal and third party services and the way the CLR handles them. The first execution of the factory delegate leads to a full evaluation, no execution cache being ready yet - once executed, it gets cached - so the next execution is slightly faster.

**Comment by trontronicent** [comment]
and PS: much thanks for this code ! :)

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 06 2019, 02:31, in reply to [Mark Vincze](https://disqus.com/by/markvincze/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/3#discussioncomment-6056278)</em>

---

<p>Yeah, this experiment pretty much leads me to the same conclusion. I'll check it out, thanks!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 06 2019, 02:30, in reply to [Muhammad Rehan Saeed](https://disqus.com/by/RehanSaeedUK/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/3#discussioncomment-6056284)</em>

---

<p>When I tested this with the health checks in 2.2, they made very little impact on the first request time. I figure that's because most of the services that are slow are related to MVC specifically, so hitting the health check endpoint doesn't help with them.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 16 2020, 12:13, in reply to [Paula Santos](https://disqus.com/by/disqus_l8VVy5GtS2/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/3#discussioncomment-6056292)</em>

---

<p>See this post for discussion on startup tasks :) <a href="https://andrewlock.net/running-async-tasks-on-app-startup-in-asp-net-core-part-2/" rel="nofollow noopener" title="https://andrewlock.net/running-async-tasks-on-app-startup-in-asp-net-core-part-2/">https://andrewlock.net/running-async-tasks-on-app-startup-in-asp-net-core-part-2/</a></p>

[a comment by an automated account is not shown]
